### PR TITLE
Update WebAuthenticator to support custom ResponseDecoder (#11641)

### LINKED ISSUE #11641: WebAuthenticator doesn't support Azure App Service callback response.
### Description

I would like to use WebAuthenticator when also using Azure App Service Authentication.  Unfortunately, EasyAuth does not return the token and expiry in a format that WebAuthenticator understands.  To fix this, I propose that we add a "ResponseDecoder" option to WebAuthenticator.

### Public API Changes

I'd suggest having a new interface `IWebAuthenticatorResponseDecoder`:

```csharp
public interface IWebAuthenticatorResponseDecoder
{
  IDictionary<string, string> DecodeResponseUri(Uri uri);
}
```

Then, in WebAuthenticator, have a new public property:

```csharp
public IWebAuthenticatorResponseDecoder? ResponseDecoder { get; set; }
```

This gets passed as an optional parameter to the constructor of `WebAuthenticatorResult` (the one that gets passed the Uri).  When the response decoder is not null, it is called to convert the Uri into a dictionary (and then everything progresses as normal).  When the response decoder is null, the current method (WebUtils.ParseQueryString()) is used to decode the response.

### Intended Use-Case

Authenticating to Azure App Service when using Azure App Service Authentication & Authorization.

The generalized method proposed allows customers to parse the callback URL for any platform, and allows App Service to release their own package that has the decoder in it if they desire.

### PR DESCRIPTION
**Pull request by adrianhall**
### Description of Change

1. Adds CallbackUri to WebAuthenticatorResult which is a read-only property.
2. Adds ResponseDecoder to the WebAuthenticatorOptions.
3. WebAuthenticatorResult uses ResponseDecoder if provided to decode the callback URL.
4. The WebAuthenticatorImplementation (iOS, Android, MacOS, UWP) stores the options when AuthenticateAsync() is called, then uses the contained ResponseDecoder to construct the WebAuthenticatorResult.
5. Device Test added to show that when the ResponseDecoder is set, the ResponseDecoder is called on callback URI.

### Issues Fixed

Fixes #11641

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/maui?pullRequest=11710) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/maui?pullRequest=11710) before we can accept your contribution.<br/><br/>:x: adrianhall [sign now](https://cla.dotnetfoundation.org/dotnet/maui?pullRequest=11710)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/maui?pullRequest=11710) it.</sub>

File: src/Essentials/src/WebAuthenticator/WebAuthenticator.shared.cs [hunk 1]
```diff
@@ -18,13 +20,19 @@ public interface IPlatformWebAuthenticatorCallback
 #endif
 	}
 
+	public interface IWebAuthenticatorResponseDecoder
+	{
+		IDictionary<string, string> DecodeResponse(Uri uri);
```
Comment (mattleibow) [inline src/Essentials/src/WebAuthenticator/WebAuthenticator.shared.cs:25]:
```suggestion
		IDictionary<string, string>? DecodeResponse(Uri uri);
```

File: src/Essentials/src/WebAuthenticator/WebAuthenticatorResult.shared.cs [hunk 2]
```diff
@@ -15,9 +15,16 @@ public WebAuthenticatorResult()
...
+		public WebAuthenticatorResult(Uri uri) : this(uri, null)
 		{
-			foreach (var kvp in WebUtils.ParseQueryString(uri.ToString()))
+		}
+
+		/// <include file="../../docs/Microsoft.Maui.Essentials/WebAuthenticatorResult.xml" path="//Member[@MemberName='.ctor'][4]/Docs/*" />
+		public WebAuthenticatorResult(Uri uri, IWebAuthenticatorResponseDecoder responseDecoder)
+		{
+			CallbackUri = uri;
+			var properties = responseDecoder == null ? WebUtils.ParseQueryString(uri.ToString()) : responseDecoder.DecodeResponse(uri);
```
Comment (mattleibow) [inline src/Essentials/src/WebAuthenticator/WebAuthenticatorResult.shared.cs:26]:
```suggestion
			var properties = responseDecoder?.DecodeResponse(uri) ?? WebUtils.ParseQueryString(uri.ToString());
```

Maybe we can do this (if my logic is correct) so if the decoder that you have provided does not support the format - for whatever reason - you can just return null and fall back to the default? Or, maybe we should throw?

If we do allow throwing, I wonder if we should have a try/catch around the constructor and pass the exception back into all the TCS?

```cs
try {
tcs.SetResult(new Result(uri, decoder));
} catch (Exception ex) {
tcs.SetException(ex);
}
```

**Review by mattleibow — COMMENTED** [review]
Thanks for this PR. I think we just made things a lot better and support more auth endpoints. And thanks again for the docs!

I only have 1 comment about the decoder results.

File: src/Essentials/src/WebAuthenticator/WebAuthenticatorResult.shared.cs (hunk 2 above)
Comment (adrianhall) [reply src/Essentials/src/WebAuthenticator/WebAuthenticatorResult.shared.cs:26]: Great points - I'll make all the changes here and do the wrapping of the exception at the same time.

**Comment by adrianhall** [comment]
... and the tcs.SetResult(new WebAuthenticatorResult(...)) was already wrapped in a SetException in each case.

File: src/Essentials/src/PublicAPI/net-android/PublicAPI.Unshipped.txt [hunk 3]
```diff
@@ -1,8 +1,14 @@
 ﻿#nullable enable
 Microsoft.Maui.ApplicationModel.Permissions.PermissionResult
+Microsoft.Maui.Authentication.IWebAuthenticatorResponseDecoder
+Microsoft.Maui.Authentication.IWebAuthenticatorResponseDecoder.DecodeResponse(System.Uri! uri) -> System.Collections.Generic.IDictionary<string!, string!>!
```
Comment (mattleibow) [inline src/Essentials/src/PublicAPI/net-android/PublicAPI.Unshipped.txt:4]:
```suggestion
Microsoft.Maui.Authentication.IWebAuthenticatorResponseDecoder.DecodeResponse(System.Uri! uri) -> System.Collections.Generic.IDictionary<string!, string!>?
```

File: src/Essentials/src/PublicAPI/net-ios/PublicAPI.Unshipped.txt [hunk 4]
```diff
@@ -1,5 +1,11 @@
 ﻿#nullable enable
+Microsoft.Maui.Authentication.IWebAuthenticatorResponseDecoder
+Microsoft.Maui.Authentication.IWebAuthenticatorResponseDecoder.DecodeResponse(System.Uri! uri) -> System.Collections.Generic.IDictionary<string!, string!>!
```
Comment (mattleibow) [inline src/Essentials/src/PublicAPI/net-ios/PublicAPI.Unshipped.txt:3]:
```suggestion
Microsoft.Maui.Authentication.IWebAuthenticatorResponseDecoder.DecodeResponse(System.Uri! uri) -> System.Collections.Generic.IDictionary<string!, string!>?
```

File: src/Essentials/src/PublicAPI/net-maccatalyst/PublicAPI.Unshipped.txt [hunk 5]
```diff
@@ -1,5 +1,11 @@
 ﻿#nullable enable
+Microsoft.Maui.Authentication.IWebAuthenticatorResponseDecoder
+Microsoft.Maui.Authentication.IWebAuthenticatorResponseDecoder.DecodeResponse(System.Uri! uri) -> System.Collections.Generic.IDictionary<string!, string!>!
```
Comment (mattleibow) [inline src/Essentials/src/PublicAPI/net-maccatalyst/PublicAPI.Unshipped.txt:3]:
```suggestion
Microsoft.Maui.Authentication.IWebAuthenticatorResponseDecoder.DecodeResponse(System.Uri! uri) -> System.Collections.Generic.IDictionary<string!, string!>?
```

File: src/Essentials/src/PublicAPI/net-tizen/PublicAPI.Unshipped.txt [hunk 6]
```diff
@@ -1,5 +1,11 @@
 ﻿#nullable enable
+Microsoft.Maui.Authentication.IWebAuthenticatorResponseDecoder
+Microsoft.Maui.Authentication.IWebAuthenticatorResponseDecoder.DecodeResponse(System.Uri! uri) -> System.Collections.Generic.IDictionary<string!, string!>!
```
Comment (mattleibow) [inline src/Essentials/src/PublicAPI/net-tizen/PublicAPI.Unshipped.txt:3]:
```suggestion
Microsoft.Maui.Authentication.IWebAuthenticatorResponseDecoder.DecodeResponse(System.Uri! uri) -> System.Collections.Generic.IDictionary<string!, string!>?
```

File: src/Essentials/src/PublicAPI/net-windows/PublicAPI.Unshipped.txt [hunk 7]
```diff
@@ -2,7 +2,13 @@
 *REMOVED*Microsoft.Maui.ApplicationModel.IWindowStateManager.ActiveWindowDisplayChanged -> System.EventHandler!
 *REMOVED*Microsoft.Maui.ApplicationModel.IWindowStateManager.OnWindowMessage(nint hWnd, uint msg, nint wParam, nint lParam) -> void
 *REMOVED*static Microsoft.Maui.ApplicationModel.Platform.OnWindowMessage(nint hWnd, uint msg, nint wParam, nint lParam) -> void
+Microsoft.Maui.Authentication.IWebAuthenticatorResponseDecoder
+Microsoft.Maui.Authentication.IWebAuthenticatorResponseDecoder.DecodeResponse(System.Uri! uri) -> System.Collections.Generic.IDictionary<string!, string!>!
```
Comment (mattleibow) [inline src/Essentials/src/PublicAPI/net-windows/PublicAPI.Unshipped.txt:6]:
```suggestion
Microsoft.Maui.Authentication.IWebAuthenticatorResponseDecoder.DecodeResponse(System.Uri! uri) -> System.Collections.Generic.IDictionary<string!, string!>?
```

File: src/Essentials/src/PublicAPI/net/PublicAPI.Unshipped.txt [hunk 8]
```diff
@@ -1,5 +1,11 @@
 ﻿#nullable enable
+Microsoft.Maui.Authentication.IWebAuthenticatorResponseDecoder
+Microsoft.Maui.Authentication.IWebAuthenticatorResponseDecoder.DecodeResponse(System.Uri! uri) -> System.Collections.Generic.IDictionary<string!, string!>!
```
Comment (mattleibow) [inline src/Essentials/src/PublicAPI/net/PublicAPI.Unshipped.txt:3]:
```suggestion
Microsoft.Maui.Authentication.IWebAuthenticatorResponseDecoder.DecodeResponse(System.Uri! uri) -> System.Collections.Generic.IDictionary<string!, string!>?
```

File: src/Essentials/src/PublicAPI/netstandard/PublicAPI.Unshipped.txt [hunk 9]
```diff
@@ -1,5 +1,11 @@
 ﻿#nullable enable
+Microsoft.Maui.Authentication.IWebAuthenticatorResponseDecoder
+Microsoft.Maui.Authentication.IWebAuthenticatorResponseDecoder.DecodeResponse(System.Uri! uri) -> System.Collections.Generic.IDictionary<string!, string!>!
```
Comment (mattleibow) [inline src/Essentials/src/PublicAPI/netstandard/PublicAPI.Unshipped.txt:3]:
```suggestion
Microsoft.Maui.Authentication.IWebAuthenticatorResponseDecoder.DecodeResponse(System.Uri! uri) -> System.Collections.Generic.IDictionary<string!, string!>?
```

**Review by mattleibow — COMMENTED** [review]
I got a sneaky feeling CI will fail with the API changes and this is going to be needed...

**Comment by jfversluis** [comment]
Hey @adrianhall thanks for your patience here! I will be monitoring this more closely now to make sure we give this the attention it deserves. Are you still up for doing this? If not, totally fine, let me know then I will take it from here :)

If yes, could you please resolve the conflicts? Also, I have been working to remove the docs from external XML files to inline docs, see: https://github.com/dotnet/maui/pulls?q=is%3Apr+is%3Aclosed+author%3Ajfversluis+label%3A%22area%2Fdocs+%F0%9F%93%93%22

So if we can fix these two things, I think the rest looks pretty good with all the feedback from Matt we had already!

**Comment by adrianhall** [comment]
Fixed merge conflicts and moved the docs into XML comments.

**Review by mattleibow — APPROVED** [review]

**Pull request merged**